smithy-model/src/main/java/software/amazon/smithy/model/transform/CreateDedicatedInputAndOutput.java

```diff
@@ -0,0 +1,220 @@
+            OperationShape operation,
+            String suffix
+    ) {
+        // Check if the shape already has the desired name.
+        ShapeId expectedName = ShapeId.fromParts(operation.getId().getNamespace(),
+                                                 operation.getId().getName() + suffix);
+        if (struct.getId().equals(expectedName)) {
+            return struct;
+        }
+
+        ShapeId newId = createSyntheticShapeId(model, operation, suffix);
+        return struct.toBuilder()
+                .id(newId)
+                .addTrait(new OriginalShapeIdTrait(struct.getId()))
+                .build();
+    }
+
+    private boolean isDedicatedHeuristic(
+            OperationShape operation,
+            StructureShape struct,
+            NeighborProvider reverse,
+            RelationshipType expected
+    ) {
+        // Only assume that a shape is dedicated to the operation its name starts with the operation name.
+        if (!struct.getId().getName().startsWith(operation.getId().getName())) {
+            return false;
+        }
+
+        // Check if the shape is only referenced as input or output.
+        return isSingularReference(reverse, struct, expected);
+    }
+
+    private boolean isSingularReference(NeighborProvider reverse, Shape shape, RelationshipType expected) {
+        int totalDirectedEdges = 0;
+
+        // We need to exclude inverted edges like MEMBER_CONTAINER, and only look for directed
+        // edges like INPUT and OUTPUT.
+        for (Relationship rel : reverse.getNeighbors(shape)) {
+            if (rel.getRelationshipType().getDirection() == RelationshipDirection.DIRECTED) {
+                totalDirectedEdges++;
+                if (rel.getRelationshipType() != expected) {
+                    return false;
+                }
+            }
+        }
+
+        return totalDirectedEdges == 1;
+    }
+
+    private static StructureShape createSyntheticShape(
+            Model model,
+            OperationShape operation,
+            String suffix,
+            StructureShape source,
+            Trait inputOutputTrait
+    ) {
+        ShapeId newId = createSyntheticShapeId(model, operation, suffix);
+        return source.toBuilder()
+                .id(newId)
+                .addTrait(inputOutputTrait)
+                .build();
+    }
+
+    private static ShapeId createSyntheticShapeId(
+            Model model,
+            OperationShape operation,
+            String suffix
+    ) {
+        // Synthesize an input shape as a dedicated copy of the existing input.
+        ShapeId newId = ShapeId.fromParts(operation.getId().getNamespace(),
+                                          operation.getId().getName() + suffix);
+
+        if (model.getShapeIds().contains(newId)) {
+            ShapeId deconflicted = resolveConflict(newId, suffix);
+            if (model.getShapeIds().contains(deconflicted)) {
+                throw new ModelTransformException(String.format(
```

Comment:
Have you run this on the AWS models to ensure it handles those cases?

Reply:
Will do

Reply:
Kinesis Analytics is a good one to test against since it has `DeleteApplication` and `DeleteApplicationOutput` operations.

Currently, smithy-rs places each generated operation input/output into its own Rust namespace so that name conflicts aren't an issue. This is illustrated in the docs for the DeleteApplication/DeleteApplicationOutput outputs in aws-sdk-kinesisanalytics:
https://docs.rs/aws-sdk-kinesisanalytics/0.0.26-alpha/aws_sdk_kinesisanalytics/output/index.html
https://docs.rs/aws-sdk-kinesisanalytics/0.0.26-alpha/aws_sdk_kinesisanalytics/operation/index.html

So if I understand these changes correctly, these would become:
- DeleteApplicationOperationOutput
- DeleteApplicationOutputOutput
Rather than:
- DeleteApplicationOutput
- DeleteApplicationOutputOutput

But the change in name is kind of moot for smithy-rs since each of these get their own Rust namespace anyway. I think I prefer leaving them as is and placing the synthetic shapes in a synthetic Smithy namespace. What do you think about making this configurable to that each code generator can decide how the conflict should be resolved?

Reply:
If you use Request/Response as your suffix, it becomes `DeleteApplicationOutputRequest` / `DeleteApplicationOutputResponse`.

And yeah, you're correct; if you use Input/Output, it becomes:
* `DeleteApplication`: `DeleteApplicationOperationOutput`
* `DeleteApplicationOutput`: `DeleteApplicationOutputOutput`

> But the change in name is kind of moot for smithy-rs since each of these get their own Rust namespace anyway. I think I prefer leaving them as is and placing the synthetic shapes in a synthetic Smithy namespace.

If you want to move input and output shapes to another namespace, then you can do that by first performing the transform, then iterating over every structure shape marked with the `@input` and `@output` shapes, and generate those into whatever namespace/package you want. You can use the applied `OriginalShapeIdTrait` if you want to further change the name (this could be added to your symbol provider for example). The transform is still useful because it will copy shapes when they're reused within inputs and outputs and it will remove unreferenced input and output shapes after copies are made.

> What do you think about making this configurable so that each code generator can decide how the conflict should be resolved?

We can't give you the exact names you have in Rust using a transform like this. Every shape, including operations, need to have unique names in the closure of a service, so you can't reuse `DeleteApplicationOutput` as both an operation name and output shape name.

Reply:
I see. Makes sense.